horse/recommendation/smart.py

```diff
@@ -0,0 +1,92 @@
+from collections import defaultdict
+
+from .base import RecommendationService
+
+
+class SmartRecommendationService(RecommendationService):
+    user_similarity_treshold = 2
+    movie_likes_fraction_exp = 0.5
+
+    def __init__(self, user_repo, movie_repo):
+        self.user_repo = user_repo
+        self.movie_repo = movie_repo
+
+    def recommend(self, user):
+        movies_to_skip = user.get_liked_movies()
+        recommendations = self._get_base_movie_recommendations(movies_to_skip)
+        user_relations = self._get_related_users(user)
+        self._add_relations_scores(
+            movies_to_skip, recommendations, user_relations)
+
+        items = sorted(recommendations.items(), key=lambda i: -i[1])
+
+        return [item[0] for item in items]
+
+    def _get_base_movie_recommendations(self, movies_to_skip):
+        all_movies = self.movie_repo.all()
+        most_likes = max([m.likes for m in all_movies])
+
+        if not most_likes:
+            return []
+
+        def calc_movie_score(movie):
+            return (movie.likes / most_likes) ** self.movie_likes_fraction_exp
+
+        movie_scores = {}
+
+        for movie in set(all_movies) - set(movies_to_skip):
+            score = calc_movie_score(movie)
+            if score:
+                movie_scores[movie] = score
+
+        return movie_scores
+
+    def _get_related_users(self, user):
+        relations = defaultdict(int)
+
+        def recur(root_user, power, depth=0):
+            if depth > 3:
```

Comment:
It would be good to have MAX_DEPTH constant defined or something similar to avoid 'magic' number.